test/xbgpu/test_engine.py

```diff
@@ -891,7 +923,12 @@ async def _send_data(
                 assert ig_recv["timestamp"].value == index * timestamp_step
                 assert ig_recv["frequency"].value == frequency
                 assert ig_recv["beam_ants"].value == np.sum(present[index])
-                beam_results[i, j, ...] = ig_recv["bf_raw"].value
+                beam_results[beam_output.name][j, ...] = ig_recv["bf_raw"].value
+
+            if beam_capture_stop_heap_index:
+                # Confirm that there are no more heaps to receive
+                with pytest.raises(spead2.Stopped):
+                    heap = await stream.get()
```

Comment:
I think this could fail to raise the Stopped exception if there is a descriptor heap that got sent just after the last data heap. If it is a descriptor heap, it should keep trying.

What I'd suggest is to take the logic from lines 867-872 (which sets and checks updated_items in a loop) and move it into a helper function, which can then be called here to (attempt to) get another data heap.